enhancements/ingress/global-admission-hsts.md

```diff
@@ -0,0 +1,568 @@
+func requiredHSTSMatchesRoute(requirement RequiredHSTSPolicy, route *routeapi.Route, namespace *corev1.Namespace) (bool, error) {
+    matchesNamespace, err := matchesNamespaceSelector(requirement.NamespaceSelector, namespace)
+    if err != nil {
+        return false, err
+    }
+    if !matchesNamespace {
+        return false, nil
+    }
+    routeDomains := []string{route.Spec.Host}
+    for _, ingress := range route.Status.Ingress {
+        routeDomains = append(routeDomains, ingress.Host)
+    }
+    if matchesDomain(requirement.DomainPatterns, routeDomains) {
+        return true, nil
+    }
+
+    return false, nil
+}
+````
+
+Formatting of the HSTS Annotation field on a `route`:
+```yaml
+max-age=<numberofseconds>[;includeSubDomains][;preload]`
+```
+
+The `RequiredHSTSPolicies` variable can be configured by the administrator in the `Ingress`, as shown in this example, where
+there is one HSTS Policy is for domain abc.com, expires after one year, includes subdomains, and requests preload:
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Ingress
+metadata:
+  name: cluster
+spec:
+  domain: apps.abc.com
+  requiredHSTSPolicies:
+    - domainPatterns:
+        - abc.com
+      maxAge:
+          smallestMaxAge: 1
+          largestMaxAge: 31536000
+      preloadPolicy: "RequirePreload"
+      includeSubdomainsPolicy: "RequireIncludeSubdomains"
+```
+Accompanying unit and e2e test changes will be added to exercise the `RequiredHSTSPolicies` type.  The API guarantees:
+* Routes with hosts that match the `domainPatterns` of the HSTS will be validated
+* Routes with hosts that match the `domainPatterns` and `namespaceSelector` of the HSTS will be validated
+* Routes that only match the `namespaceSelector` will not be validated
+* Routes that are validated will validate that the `maxAge` exists and falls within the range of the HSTS `maxAgePolicy`
+* Routes that are validated will validate that the `preloadPolicy` and `includeSubdomainPolicy` match the HSTS Policy if they exist
+
+#### Audit HSTS configurations on any cluster
+This proposal offers the cluster administrator a way to review the HSTS configurations guaranteed by the admission plugin.
+For example, to review the maxAge set for HSTS:
+```shell
+oc describe ingresscontroller default -n openshift-ingress-controller -o jsonpath='{range .items[*].spec.requiredHSTSPolicies}{"Max Age: "}{.largestMaxAge}{"\n"}{end}'
+```
+
+#### Predict HSTS enforcement on any cluster with a single configuration manifest
+This proposal offers the cluster administrator a way to configure and enforce HSTS configurations on routes by configuring
+the `Ingress` using templates that substitute only the for the difference between clusters.  For example, for
+a development cluster with domain `dev.abc.com` and production cluster with `prod.abc.com` domains, administrators can use a single
+`Ingress` deployment template that substitutes the domain:
```

Comment:
```suggestion
`Ingress` config template that substitutes the domain:
```
I think you mean "Ingress Config" instead of "Ingress Deployment" in a few places in this enhancement?